src/webgpu/util/math.ts

```diff
@@ -621,57 +625,136 @@ export function correctlyRoundedF16(n: number): number[] {
   return [Number.NEGATIVE_INFINITY];
 }
 
-/**
- * Once-allocated ArrayBuffer/views to avoid overhead of allocation in frexp
- *
- * This makes frexp non-reentrant due to shared state between calls.
- */
-const frexpData = new ArrayBuffer(4);
-const frexpDataU32 = new Uint32Array(frexpData);
-const frexpDataF32 = new Float32Array(frexpData);
-
 /**
  * Calculates WGSL frexp
  *
  * Splits val into a fraction and an exponent so that
  * val = fraction * 2 ^ exponent.
  * The fraction is 0.0 or its magnitude is in the range [0.5, 1.0).
  *
- * Inspired by golang's implementation of frexp.
- *
- * This code is non-reentrant due to the use of a non-local data buffer and
- * views.
- *
- * @param val the f32 to split
+ * @param val the float to split
+ * @param trait the float type, f32 or f16 or f64
  * @returns the results of splitting val
  */
-export function frexp(val: number): { fract: number; exp: number } {
-  frexpDataF32[0] = val;
-  // Do not directly use val after this point, so that changes are reflected in
-  // both the f32 and u32 views.
+export function frexp(val: number, trait: 'f32' | 'f16' | 'f64'): { fract: number; exp: number } {
```

Comment:
This function can be simplified by always handling the `f64` trait i.e. JS `number`, with input value quantized to different types (`f32` or `f16`). The result should be exactly the same, and the input quantizing is already used in `frexp.spec.ts`.